File: my_modules/queries.js
var queries = function() {

  var self = this;

  this.getAllInfo = function() {
    var locations = self.getAllLocations();
    var festivals = self.getAllFestivals();
    var artists = self.getAllArtists();
    var festivalDates = self.getAllFestivalDates();
    var appearances = self.getAllAppearances();
    var chordData = self.getChordData();
    var sql = locations.cmd.sql +
      festivals.cmd.sql +
      artists.cmd.sql +
      festivalDates.cmd.sql +
      appearances.cmd.sql +
      chordData.cmd.sql;
    return {
      cmd: { sql: sql },
      process: function(dbRes) {
        var locObj = locations.process(dbRes[0]);
        var festObj = festivals.process(dbRes[1]);
        var artObj = artists.process(dbRes[2]);
        var festDateObj = festivalDates.process(dbRes[3]);
        var appObj = appearances.process(dbRes[4]);
        var chordObj = chordData.process(dbRes[5]);
        return {
          locations: locObj,
          festivals: festObj,
          artists: artObj,
          festivalDates: festDateObj,
          appearances: appObj,
          chordData: chordObj
        };
      }
    };
  };

  this.getAllLocations = function() {
    return {
      cmd: { sql: 'select * from locations;' },
      process: function(dbRes) {
        return dbRes.map(function(v,i){
          return [v.locationId,v.city,v.state,v.country,parseFloat(v.lat),parseFloat(v.lng)];
        });
      }
    };
  };

  this.getAllFestivals = function(orderBy) {
    return {
      cmd: { sql: 'select * from festivals;' },
      process: function(dbRes) {
        return dbRes.map(function(v,i){
          return [v.festivalId,v.festival,v.locationId,v.website];
        });
      }
    };
  };

  this.getAllArtists = function() {
    var sql = 'select * from artists ar where exists ' +
      '(select 1 from appearances ap where ap.artistId = ar.artistId) ' +
      'and not exists (select 1 from artistBlacklist ab where ab.artistId = ar.artistId);';
    return {
      cmd: { sql: sql },
      process: function(dbRes) {
        return dbRes.map(function(v,i){
          return [v.artistId,v.artist];
        });
      }
    };
  };

  this.getAllFestivalDates = function(orderBy) {
    return {
      cmd: { sql: 'select * from festivalDates;' },
      process: function(dbRes) {
        return dbRes.map(function(v,i){
          return [v.festivalDateId,v.festivalId,v.week,v.startDate,v.endDate];
        });
      }
    };
  };

  this.getAllAppearances = function() {
    var sql = 'select * from appearances ap where not exists ' +
      '(select 1 from artistBlacklist ab where ab.artistId = ap.artistId);';
    return {
      cmd: { sql: sql },
      process: function(dbRes) {
        return dbRes.map(function(v,i){
          return [v.festivalDateId,v.artistId];
        });
      }
    };
  };

  this.getChordData = function() {
    var sql = '\
select \
 f1.festivalId as festivalId1 \
,f2.festivalId as festivalId2 \
,coalesce(c.count,0) as count \
from festivals f1 \
 cross join festivals f2 \
 left outer join \
  (select \
    a1.festivalId as festivalId1 \
  , a2.festivalId as festivalId2 \
  , count(1) as count \
  from (select distinct \
        artistId \
        ,festivalId \
        from appearances a \
         inner join festivalDates fd \
          on fd.festivalDateId = a.festivalDateId) a1 \
   inner join (select distinct \
               artistId \
               ,festivalId \
               from appearances a \
                inner join festivalDates fd \
                 on fd.festivalDateId = a.festivalDateId) a2 \
    on a2.artistId = a1.artistId \
  where a1.festivalId <> a2.festivalId \
  group by a1.festivalId,a2.festivalId) c \
  on c.festivalId1 = f1.festivalId \
  and c.festivalId2 = f2.festivalId \
where f1.festivalId <= f2.festivalId \
order by f1.festivalId,f2.festivalId;';
    return {
      cmd: { sql: sql },
      process: function(dbRes) {
        return dbRes.map(function(v,i){
          return [v.festivalId1,v.festivalId2,v.count];
        });
      }
    };
  };

  this.getBlacklist = function() {
    var sql = '\
select \
  a.artistId \
, a.artist \
, case \
  when exists (select 1 from artistBlacklist ab where ab.artistId = a.artistId) \
    then true \
  else false \
  end as blacklisted \
, group_concat(f.festival SEPARATOR \', \') as festivals \
from artists a \
  inner join appearances ap \
    on ap.artistId = a.artistId \
  inner join festivalDates fd \
    on fd.festivalDateId = ap.festivalDateId \
  inner join festivals f \
    on f.festivalId = fd.festivalId \
group by a.artistId,a.artist \
order by artist asc;';
    return { sql: sql };
  };

  this.addBlacklist = function(artistId) {
    var sql = '\
insert into artistBlacklist(artistId) \
select t.artistId from (select ? as `artistId`) t \
where not exists \
(select 1 from artistBlacklist ab where ab.artistId = t.artistId);'
    return {
      sql: sql,
      inserts: [artistId,artistId]
    };
  };

  this.unBlacklist = function(artistId) {
    return {
      sql: 'delete from artistBlacklist where artistId = ?;',
      inserts: [artistId]
    };
  };

};

module.exports = new queries();

File: py/queries.py
def getAllArtists():
  return 'select * from artists;'

def getArtistInfo(artistId):
  return ('Select '
    'ap.appearanceId, '
    'ap.setTime, '
    'fest.festivalId, '
    'fest.festival, '
    'fest.week, '
    'fest.location, '
    'fest.startDate, '
    'fest.endDate '
    'from appearances ap '
    '  inner join festivals fest '
    '    on fest.festivalId = ap.festivalId '
    'where ap.artistId = ' + str(artistId) + ' '
    'order by fest.startDate asc;')

def getFestivalsForArtist(artistId):
  return ('Select '
    'GROUP_CONCAT(ap.appearanceId separator \',\') as appearanceIds, '
    'GROUP_CONCAT(ap.setTime separator \',\') as setTimes, '
    'GROUP_CONCAT(fest.festivalId separator \',\') as festivalIds, '
    'fest.festival, '
    'GROUP_CONCAT(fest.week separator \',\') as weeks, '
    'GROUP_CONCAT(distinct fest.location separator \',\') as location, '
    'GROUP_CONCAT(fest.startDate separator \',\') as startDates, '
    'GROUP_CONCAT(fest.endDate separator \',\') as endDates '
    'from appearances ap '
      'inner join festivals fest '
        'on fest.festivalId = ap.festivalId '
    'where ap.artistId = ' + str(artistId) + ' '
    'group by festival '
    'order by startDate;')

  this.getUniqueFestivals = function() {
    return 'select * from festivals where festivalId <> 2;';
  };

  this.getFestival = function(festivalId) {
    return 'select * from festivals where festivalId = ' + festivalId + ';';
  };

  this.getFestivalInfo = function(festivalId) {
    return {
      festival: this.getFestival(festivalId),
      artists: this.getArtistsForFestival(festivalId)
    };
  };

  this.getAllFestivals = function(orderBy) {
    return 'select * from festivals ' + 
      (typeof orderBy !== 'undefined' ? 
        'order by ' + orderBy + ' asc' : '') + ';';
  };

  this.getArtistsForFestival = function(festivalId) {
    return 'select \
  ar.artistId, \
  ar.artist, \
  app.appearanceId, \
  app.setTime \
from artists ar \
  inner join appearances app \
    on app.artistId = ar.artistId \
where app.festivalId = ' + festivalId + 
' order by artist asc;';
  };

  this.getOrderedFestivals = function(festivalIds) {
    return 'Select \
              festivalId, \
              festival, \
              @curRow := @curRow + 1 AS idx \
            from festivals f \
            join (select @curRow := 0) r \
            where festivalId in (' + festivalIds + ') \
            order by festivalId asc;';
  };

  this.getOverlapsForFestivals = function(festivals) {
    var data = [];
    var base = '(select cast(count(1) as char(10)) from appearances a';
    var append = ' \n and exists (select 1 from appearances a2 \
      \n where a1.artistId = a2.artistId and a2.festivalId = ';

    var buildBase = function(festival) {
      return {
        sql: '(select cast(count(1) as char(10)) from appearances a' + festival.festivalId + 
          ' where festivalId = ' + festival.festivalId,
        fests: [festival]
      };
    };

    var appendBase = function(baseObj,festival) {
      var a = baseObj.fests.slice(0);
      a.push(festival);
      return {
        sql: baseObj.sql + ' \n and exists (select 1 from appearances a' +
          festival.festivalId + '  where a' + baseObj.fests[0].festivalId + 
          '.artistId = a' + festival.festivalId + '.artistId' +
          ' and a' + festival.festivalId + '.festivalId = ' + 
          festival.festivalId + ')',
        fests: a
      };
    }

    for (var i = 0; i < festivals.length; i++) {
      var d_length = data.length;
      for (var j = 0; j < d_length; j++) {
        data.push(appendBase(data[j],festivals[i]));
      }
      data.push(buildBase(festivals[i]));
    }

    var setBase = 'select CONCAT(\'[\',';
    var setArray = [];
    var overlapBase = 'select CONCAT(\'[\',';
    var overlapArray = [];
    for (var i = 0; i < data.length; i++) {
      if (data[i].fests.length > 1) {
        var tag = '[';
        var overlap = '';
        if (overlapArray.length > 0) {
          overlap += '),\'}, \',';
        }
        for (var j = 0; j < data[i].fests.length; j++) {
          tag += data[i].fests[j].idx - 1;
          if (j + 1 == data[i].fests.length) {
            tag += ']';  
          } else {
            tag += ','
          }
        }
        overlap += '\'{"sets": ' + tag + ', "size": \',' + data[i].sql;
        overlapArray.push(overlap);
      } else {
        var set = '';
        if (setArray.length > 0) {
          set += '),\'}, \',';
        }
        set += '\'{"label": "' + data[i].fests[0].festival.replace('\'','\\\'') + 
          '", "festivalId": "' + data[i].fests[0].festivalId + '", "size": \',' + data[i].sql;
        setArray.push(set);
      }
    }
    var sets = setBase + setArray.join('\n') + '),\'}]\') as sets;';
    var overlaps = overlapBase + overlapArray.join('\n') + '),\'}]\') as overlaps;';
    return {
      sets: sets,
      overlaps: overlaps
    };
  };

  this.getInCommonForFestivals = function(festivalIds) {
    var festivals = festivalIds.split(',');
    var query = 'Select ar.* from artists ar';

    var appendBase = function(id,index) {
      return ' \n ' + (index === 0 ? 'where' : 'and') + 
        ' exists (select 1 from appearances a' + id + 
        ' \n  where ar.artistId = a' + id + '.artistId' +
        ' and a' + id + '.festivalId = ' + id + ')';
    }

    for (var i = 0; i < festivals.length; i++) {
      query += appendBase(festivals[i],i);
    }

    query += ' order by ar.artist;';
    return query;
  };
};

module.exports = new queries();